spharpy/spherical.py

```diff
@@ -1,104 +1,585 @@
+            The default is True.
+        """
+        if not isinstance(coords, pf.Coordinates):
+            raise TypeError("coords must be a pyfar.Coordinates object")
+        self.n_max = n_max
+        self.weights = weights
+        self.coords = coords
+        self.basis_type = basis_type
+        self.inverse_transform = inverse_transform
+        self.channel_convention = channel_convention  # ch. ord. conv.
+        self.normalization = normalization  # gain norm. conv.
+        self.basis = None
+        (self.basis_gradient_theta,
+         self.basis_gradient_phi) = (None, None)
+        self._basis_inv = None
+        (self._basis_inv_gradient_theta,
+         self._basis_inv_gradient_phi) = (None, None)
+        self._condon_shortley = condon_shortley
+        # Store previous values for comparison
+        self._prev_n_max = self.n_max
+        self._prev_coords = self.coords
+        self._prev_basis_type = self.basis_type
+        self._prev_inverse_transform = self.inverse_transform
+        self._prev_channel_convention = self.channel_convention
+        self._prev_normalization = self.normalization
+        self._prev_condon_shortley = condon_shortley
+
+    # Properties
+    @property
+    def n_max(self):
+        return self._n_max
+
+    @n_max.setter
+    def n_max(self, value):
+        if value < 0:
+            raise ValueError("n_max must be a positive integer")
+        if value % 1 != 0:
+            raise ValueError("n_max must be an integer value")
+        self._n_max = int(value)  # Cast to int for safety
+
+    @property
+    def weights(self):
+        """Sampling weights for numeric integration."""
+        return self._weights
+
+    @weights.setter
+    def weights(self, weights):
+        if weights is None:
+            self._weights = None
+            return
+        if len(weights) != len(self.coords.cartesian):
+            raise ValueError(
+                "The number of weights has to be equal to \
+                    the number of sampling points."
+            )
+
+        weights = np.asarray(weights, dtype=float)
+        norm = np.linalg.norm(weights, axis=-1)
+
+        if not np.allclose(norm, 4 * np.pi):
+            weights *= 4 * np.pi / norm
+
+        self._weights = weights
+
+    @property
+    def coords(self):
+        return self._coords
+
+    @coords.setter
+    def coords(self, value):
+        if not isinstance(value, pf.Coordinates):
+            raise TypeError("coords must be a pyfar.Coordinates object")
+        if value.cdim != 1:
+            raise ValueError("Coordinates must be 1D")
+            # TODO: Allow for 2D coordinates
+        self._coords = value
+
+    @property
+    def basis_type(self):
+        return self._basis_type
+
+    @basis_type.setter
+    def basis_type(self, value):
+        assert value in ["complex", "real"], (
+            "Invalid basis type, currently only 'complex' and"
+            " 'real' "
+            "are supported"
+        )
```

Comment:
This raises an AssertionError, which is generally not very user friendly. We typically raise ValueErrors in cases of faulty user input.